chore(build): Add scripts to build images using github action

benoitf opened the pull request:
### What does this PR do?

Add a script to build either all images, a given set of images or only the images changed in the last commit

```
./build.sh --help
./build.sh : will build only modified folders since the last commit
./build.sh --all-images: rebuild all images : will build only modified folders since the last commit
./build.sh --folders=path1,path2: rebuild only those paths
```


### What issues does this PR fix or reference?
https://issues.redhat.com/browse/CRW-2081

### How to test this PR?
Log in to registry.redhat.io and then:

```
./.github/build.sh --folders=codeready-workspaces-traefik
```

nickboldt commented on .ci/build.sh:
```diff
@@ -0,0 +1,230 @@
+    local -r dockerfilePath=${1}
+
+    # Patch Dockerfile to add redhat registry
+    sed "s/FROM rhel8/FROM registry.redhat.io\/rhel8/" "${dockerfilePath}" | \
```
this might not work -- some images use public ubi8 org, but internally use rhel8 org for container references. 

So... might need more complicated rules here for the go-toolset base images.

benoitf replied:
could you be more specific ?

for example machine-exec is using `FROM rhel8/go-toolset:1.15.13-4 as builder` and it's replaced by `FROM registry.redhat.io/rhel8/go-toolset:1.15.13-4 as builder` and then it works

nickboldt commented:
I see two "fixme" items in there -- can you resolve those before merging?

Also, I see we're using .github for the action and .ci for the script -- would it make sense to have everything in the .github folder, since the .ci folder convention seems to be going away ... we don't use it in Che any more for centos CI stuff, and we've renamed builds from ":ci" to ":next" ...

benoitf commented:
> Also, I see we're using .github for the action and .ci for the script -- would it make sense to have everything in the .github folder, since the .ci folder convention seems to be going away ... we don't use it in Che any more for centos CI stuff, and we've renamed builds from ":ci" to ":next" ...

I was wondering if we could have the script at the root folder.
As there is nothing related to github action in this file. but let's keep it for now 'internally in github folder" 👍

benoitf commented:
> I see two "fixme" items in there -- can you resolve those before merging?
One is an enhancement, I've renamed fixme

the other fixme is related to  https://issues.redhat.com/browse/CRW-2077 so I added a link to it

benoitf commented:
it should be good for step 1 now :-)

nickboldt commented on .github/build.sh:
```diff
@@ -0,0 +1,231 @@
+        help
+        exit 0;;
+      *)
+        FILTERED_ARGS="${FILTERED_ARGS} $1"
```
why are you collecting unused args? doesn't seem like you're doing anything with these. I'm sure the shellcheck gods would be angry about this. :)

benoitf replied:
shellcheck is passing. You can verify on shellcheck.net

benoitf replied:
(in any case I've removed the variable but it was not complaining)

nickboldt commented on .github/build.sh:
```diff
@@ -0,0 +1,231 @@
+get_current_version() {
+  # get the current version from the VERSION file of codeready-workspaces
+  # could look at a file inside the repository once https://issues.redhat.com/browse/CRW-2077 is done
+  curl -s https://raw.githubusercontent.com/redhat-developer/codeready-workspaces/HEAD/dependencies/VERSION
```
I generally curl with -sSL in case of errors and links/redirections... also does HEAD always map to the default branch's latest commit? If so, this is a problem since I imagine at some point you'll want to run this script from the crw-2.y-rhel-8 stable branch too, not just the nightly/CI/next branch crw-2-rhel-8. 

Would it be better to compute the correct VERSION file from the current branch, eg., with `git rev-parse --abbrev-ref HEAD` ?

benoitf replied:
the would it better is the fix of https://issues.redhat.com/browse/CRW-2077 (the VERSION file in this repository)
I would not investigate more there.

And as it's only applied on default branch, HEAD is working.

nickboldt commented on .github/build.sh:
```diff
@@ -0,0 +1,231 @@
+    search_filename=${filenameWithoutExtension}@${expectedSha512}${fileExtension}
+
+    # find the matching asset
+    jq --arg SEARCH_FILENAME "${search_filename}" -r '. | select(.name==$SEARCH_FILENAME) | .browser_download_url' <<< "${assetJsonContent}"
```
this likely needs a shellcheck disable on it, so it doesn't complain about `$SEARCH_FILENAME` not being in ""

benoitf replied:
I don't have any shellcheck reports so what shellcheck are you using ?

benoitf replied:
$SEARCH_FILENAME is a jq variable and we properly give the assignement in `--arg SEARCH_FILENAME "${search_filename}"`

and all jq parameter is wrapped into simple quotes to not interpret any variable so is it you that is suspicious but not shellcheck ?

nickboldt commented on .github/build.sh:
```diff
@@ -0,0 +1,231 @@
+    # Patch Dockerfile to add redhat registry
+    sed "s/FROM rhel8/FROM registry.redhat.io\/rhel8/" "${dockerfilePath}" | \
+    sed "s/FROM ubi8/FROM registry.redhat.io\/ubi8/"
+
```
empty line

nickboldt commented on .github/build.sh:
```diff
@@ -0,0 +1,231 @@
+# analyze args
+parse "$@"
+
+# build images, with either:
```
to avoid writing the same instructions/usage in 2 places, why not just mention the `parse()` and `help()` methods above?

nickboldt replied:
also you don't mention that we can run this with no params and it should look for recent changes to build.

benoitf replied:
> also you don't mention that we can run this with no params and it should look for recent changes to build.

it's written there: only images that are modified

benoitf replied:
deleted comments and added a see usage()

nickboldt commented on .github/workflows/build-image.yaml:
```diff
@@ -0,0 +1,50 @@
+  workflow_dispatch:
+    inputs:
+      arguments:
+        description: 'CLI Argument'     
```
A better description might be "parameters for the build" with examples like 
```
--all-images or --folders=codeready-workspaces-machineexec,codeready-workspaces-imagepuller,...
```

nickboldt commented on .github/workflows/build-image.yaml:
```diff
@@ -0,0 +1,50 @@
+      with:
+        fetch-depth: 0
+
+#    - name: Login to quay.io
```
Does this need a `TODO: enable this when ready to push images to quay` ?

nickboldt commented on .github/build.sh:
```diff
@@ -0,0 +1,231 @@
+    local imageName="quay.io/crw/${folderToBuild}:gh-${currentVersion}"
+    echo "${imageBuildCount} - Building image ${imageName} in the folder ${folderToBuild}..."
+    build_image "${folderToBuild}" "${imageName}" "${availableAssets}"
+    echo "Image ${imageName} successfully build."
```
s/build/built/

Also, is this where a `TODO: push multiarch images to quay.io/crw/...` should be?

benoitf replied:
well for now there is no push so we don't need TODO.
it requires all assets to be there before doing the next step.

nickboldt commented on .github/build.sh:
```diff
@@ -0,0 +1,231 @@
+  
+  # if foldersToBuild is empty
+  if [ -z "${foldersToBuild}" ]; then
+    echo " => No image to build, skipping"
```
error message should clarify that none of the 3 conditions have been met:
* all images flag
* list of specific images via flag
* local/recent commits

benoitf replied:
added a 'mode' echo before

nickboldt commented on .github/workflows/build-image.yaml:
```diff
@@ -0,0 +1,50 @@
+      uses: docker/login-action@v1
+      with:
+        registry: registry.redhat.io
+        username: ${{ secrets.REGISTRY_REDHAT_IO_USERNAME }}
```
dumb question... did you already upload the crw-bot secret so this will work?

benoitf replied:
yes and it works (I tested in my fork as well)

nickboldt commented:
TODO before merging:
* generate VERSION file in images repo
* publish via automated process all the assets currently in pkgs.devel to GH, so that this process can be enabled

benoitf commented:
Updated the script as sha512 are not used anymore

it's still missing the CRW version in this repository / https://issues.redhat.com/browse/CRW-2077

Tested on traefik folder/x64

nickboldt commented:
Blocked by:
* GH release asset IDs should match their container names - https://issues.redhat.com/browse/CRW-2436
* CSV versions in job-config.json are invalid for newly added versions (2.14) and for 2.x - https://issues.redhat.com/browse/CRW-2435
* Add version of CodeReady workspaces in codeready-workspaces-images repository - https://issues.redhat.com/browse/CRW-2077

benoitf commented:
https://issues.redhat.com/browse/CRW-2436 is not a blocker (I have filtered out `stacks` and `plugin`) so AFAIK it could be merged

nickboldt commented on .github/build.sh:
```diff
@@ -0,0 +1,226 @@
+    local -r dockerfilePath=${1}
+
+    # Patch Dockerfile to add redhat registry
+    sed "s/FROM rhel8/FROM registry.redhat.io\/rhel8/" "${dockerfilePath}" | \
```
this could be done as a single sed and actually save the change to the file...
```
sed -r -i "${dockerfilePath}" -e  "s#FROM (rhel8|ubi8)#FROM registry.redhat.io/(\1)#g"
```

benoitf replied:
yes but -i is not macos friendly

nickboldt replied:
that's why you wrote sed_in_place for theia, right? :)
